File: mr_reviewer/parallel.py
import logging
from concurrent.futures import ThreadPoolExecutor, as_completed

from mr_reviewer.exceptions import ProviderError
from mr_reviewer.models import DiffFile, MRMetadata, ReviewComment, ReviewResult
from mr_reviewer.diff_parser import annotate_diff
from mr_reviewer.prompts import build_system_prompt, build_user_message
from mr_reviewer.providers.base import ReviewProvider

logger = logging.getLogger(__name__)


def _build_partial_diff(diff_files: list[DiffFile]) -> str:
    """Build a unified diff string from a subset of DiffFiles."""
    from mr_reviewer.core import build_unified_diff  # noqa: PLC0415
    return build_unified_diff(diff_files)


def _merge_results(results: list[ReviewResult]) -> ReviewResult:
    """Merge multiple ReviewResults into one.

    - Concatenates summaries with separator
    - Deduplicates comments by (file, line) — first occurrence wins
    - Sorts merged comments by (file, line)
    """
    if not results:
        return ReviewResult(summary="", comments=[])

    summaries = [r.summary for r in results if r.summary]
    merged_summary = "\n\n---\n\n".join(summaries)

    seen: set[tuple[str, int]] = set()
    merged_comments: list[ReviewComment] = []
    for result in results:
        for comment in result.comments:
            key = (comment.file, comment.line)
            if key not in seen:
                seen.add(key)
                merged_comments.append(comment)

    merged_comments.sort(key=lambda c: (c.file, c.line))

    return ReviewResult(summary=merged_summary, comments=merged_comments)


def parallel_review(
    provider: ReviewProvider,
    diff_files: list[DiffFile],
    file_contents: dict[str, str],
    focus_areas: list[str],
    metadata: MRMetadata,
    num_agents: int = 2,
    max_comments: int = 10,
) -> ReviewResult:
    """Run parallel reviews by partitioning diff_files across num_agents.

    Files are distributed round-robin across agents. Each agent reviews its
    partition concurrently. Results are merged with deduplication.

    If some agents fail, partial results are returned. If all agents fail,
    ProviderError is raised.
    """
    # Partition diff_files round-robin across agents
    partitions: list[list[DiffFile]] = [[] for _ in range(num_agents)]
    for i, df in enumerate(diff_files):
        partitions[i % num_agents].append(df)

    system_prompt = build_system_prompt(focus_areas, max_comments=max_comments)

    def run_agent(partition: list[DiffFile]) -> ReviewResult:
        partial_diff = _build_partial_diff(partition)
        # Only include file_contents for files in this partition
        partition_paths = {df.new_path for df in partition} | {df.old_path for df in partition}
        partial_contents = {
            path: content
            for path, content in file_contents.items()
            if path in partition_paths
        }
        user_message = build_user_message(
            title=metadata.title,
            description=metadata.description,
            diff=annotate_diff(partial_diff),
            file_contents=partial_contents,
        )
        return provider.run_review(system_prompt, user_message)

    results: list[ReviewResult] = []
    futures = {}

    with ThreadPoolExecutor(max_workers=num_agents) as executor:
        for i, partition in enumerate(partitions):
            if not partition:
                continue
            future = executor.submit(run_agent, partition)
            futures[future] = i

        for future in as_completed(futures):
            agent_idx = futures[future]
            try:
                result = future.result()
                results.append(result)
                logger.info("Agent %d completed review successfully", agent_idx)
            except Exception as e:
                logger.error("Agent %d failed: %s", agent_idx, e)

    if not results:
        raise ProviderError("All parallel review agents failed")

    return _merge_results(results)

File: internal/review/parallel.go
package review

import (
	"context"
	"fmt"
	"sort"
	"sync"
)

const defaultParallelAgents = 2

// ParallelReview partitions files round-robin across agents, reviews concurrently,
// and merges results. Partial agent failure is ignored; all-fail returns an error.
func ParallelReview(ctx context.Context, provider Provider, files []DiffFile, contents map[string]string, focus []string, meta Metadata, agents, maxComments int) (Result, error) {
	if provider == nil {
		return Result{}, fmt.Errorf("review provider is required")
	}
	if agents <= 0 {
		agents = defaultParallelAgents
	}
	if maxComments <= 0 {
		maxComments = 10
	}
	if len(focus) == 0 {
		focus = DefaultFocus
	}

	partitions := make([][]DiffFile, agents)
	for i, df := range files {
		partitions[i%agents] = append(partitions[i%agents], df)
	}

	system := SystemPrompt(focus, maxComments)
	type outcome struct {
		result Result
		err    error
	}
	var (
		mu      sync.Mutex
		results []Result
		wg      sync.WaitGroup
	)
	for _, part := range partitions {
		if len(part) == 0 {
			continue
		}
		part := part
		wg.Add(1)
		go func() {
			defer wg.Done()
			select {
			case <-ctx.Done():
				return
			default:
			}
			out, err := reviewPartition(ctx, provider, system, part, contents, meta)
			if err != nil {
				return
			}
			mu.Lock()
			results = append(results, out)
			mu.Unlock()
		}()
	}
	wg.Wait()
	if ctx.Err() != nil {
		return Result{}, ctx.Err()
	}
	if len(results) == 0 {
		return Result{}, fmt.Errorf("all parallel review agents failed")
	}
	return MergeResults(results), nil
}

func reviewPartition(ctx context.Context, provider Provider, system string, files []DiffFile, contents map[string]string, meta Metadata) (Result, error) {
	paths := map[string]struct{}{}
	for _, df := range files {
		if df.NewPath != "" {
			paths[df.NewPath] = struct{}{}
		}
		if df.OldPath != "" {
			paths[df.OldPath] = struct{}{}
		}
	}
	partial := map[string]string{}
	for path, content := range contents {
		if _, ok := paths[path]; ok {
			partial[path] = content
		}
	}
	user := UserMessage(meta.Title, meta.Description, AnnotateDiff(BuildUnifiedDiff(files)), partial)
	return provider.Review(ctx, system, user)
}

// MergeResults concatenates summaries, keeps the first comment per (file, line), and sorts.
func MergeResults(results []Result) Result {
	if len(results) == 0 {
		return Result{Summary: "", Comments: []Comment{}}
	}
	var summaries []string
	seen := map[string]struct{}{}
	var comments []Comment
	for _, r := range results {
		if r.Summary != "" {
			summaries = append(summaries, r.Summary)
		}
		for _, c := range r.Comments {
			key := fmt.Sprintf("%s\x00%d", c.File, c.Line)
			if _, ok := seen[key]; ok {
				continue
			}
			seen[key] = struct{}{}
			comments = append(comments, c)
		}
	}
	sort.SliceStable(comments, func(i, j int) bool {
		if comments[i].File != comments[j].File {
			return comments[i].File < comments[j].File
		}
		return comments[i].Line < comments[j].Line
	})
	if comments == nil {
		comments = []Comment{}
	}
	summary := ""
	for i, s := range summaries {
		if i > 0 {
			summary += "\n\n---\n\n"
		}
		summary += s
	}
	return Result{Summary: summary, Comments: comments}
}
